File: backend/app/routers/admin/reports.py
from typing import Literal

from fastapi import APIRouter, Depends
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import AdminClaims, require_super_admin
from app.db import get_session
from app.models import Booking, BookingStatus

router = APIRouter(prefix="/api/admin/reports", tags=["admin-reports"])


@router.get("")
async def reports(
    type: Literal["revenue", "source", "status"] = "revenue",
    session: AsyncSession = Depends(get_session),
    admin: AdminClaims = Depends(require_super_admin),
):
    if type == "revenue":
        result = await session.execute(
            select(Booking.property_id, func.sum(Booking.total_amount))
            .where(Booking.status == BookingStatus.confirmed)
            .group_by(Booking.property_id)
        )
    elif type == "source":
        result = await session.execute(
            select(Booking.source, func.count(Booking.id)).group_by(Booking.source)
        )
    else:
        result = await session.execute(
            select(Booking.status, func.count(Booking.id)).group_by(Booking.status)
        )
    return [{"key": str(row[0]), "value": row[1]} for row in result.all()]

File: backend/src/routes/admin/reports.ts
import { Router } from "express";
import { query } from "../../db";
import { asyncHandler, HttpError } from "../../middleware/errors";
import { requireSuperAdmin } from "../../security";
import { BookingStatus } from "../../types/enums";

export const router = Router();

router.use(requireSuperAdmin);

router.get(
  "",
  asyncHandler(async (req, res) => {
    const type = String(req.query.type ?? "revenue");
    if (!["revenue", "source", "status"].includes(type)) throw new HttpError(422, "Invalid report type");

    let rows: { key: unknown; value: unknown }[];
    if (type === "revenue") {
      rows = await query(
        `SELECT property_id AS key, SUM(total_amount) AS value FROM bookings WHERE status = $1 GROUP BY property_id`,
        [BookingStatus.confirmed]
      );
    } else if (type === "source") {
      rows = await query(`SELECT source AS key, COUNT(*) AS value FROM bookings GROUP BY source`);
    } else {
      rows = await query(`SELECT status AS key, COUNT(*) AS value FROM bookings GROUP BY status`);
    }

    res.json(rows.map((row) => ({ key: String(row.key), value: typeof row.value === "string" ? parseInt(row.value, 10) : row.value })));
  })
);
